### PR TITLE
SN-3319 NMEA additions

### PR DESCRIPTION
**Pull request by waltjohnson**
Add NMEA GSV (satellite info) message.  Fixed issue that prevents PASHR message from streaming.  Fixed other odd NMEA message bugs via unit and ci-hdw tests.  Added unit tests for all NMEA messages.  Fixed ci-hdw tests for GSA, ZDA, and RMC.  Added ci-hdw test for ASCB and ASCE.  Created NMEA stream enable control message ASCE which functionally replaces ASCB and abstracts control of NMEA message by using a key value pair: NMEA ID (eNmeaAsciiMsgId) and period.  Creates DID_GPSx_SIG message to convey multi-band signal information, required for NMEA GSV support. 

These changes are motivated for the GPX NMEA release but were made on the development branch to make use of the functional and ci-hdw supported code base.

### CODE REVIEW

File: src/protocol_nmea.cpp [hunk 1]
```diff
@@ -303,217 +418,213 @@ int did_dev_info_to_nmea_info(char a[], const int aSize, dev_info_t &info)
...
-	int n = SNPRINTF(a, aSize, "$PTOW");
-	n += SNPRINTF(a+n, aSize-n, ",%.6lf", imuTow);			// 1
-	n += SNPRINTF(a+n, aSize-n, ",%.6lf", insTow);			// 2
-	n += SNPRINTF(a+n, aSize-n, ",%u", gpsWeek);			// 3
-	
-	unsigned int checkSum = ASCII_compute_checksum((uint8_t*)(a+1), n);
-	n += SNPRINTF(a+n, aSize-n, "*%.2x\r\n", checkSum);
-	return n;	
+	int n = ssnprintf(a, aSize, "$PTOW");
+	nmea_sprint(a, aSize, n, ",%.6lf", imuTow);			// 1
```
Comment (kylemallory) [inline src/protocol_nmea.cpp:427]: I like the new use of nmea_sprint() though I'd probably use `nmea_sprintf` to be clear that its a printf derivative.  But, I'm not sure the meaning of the  inline-comments numbering // 1, 2, 3, etc...  Either these should go away, or there should be some supporting comments to communicate their purpose.

File: src/protocol_nmea.cpp [hunk 2]
```diff
@@ -851,11 +870,463 @@ int did_gps_to_nmea_pashr(char a[], const int aSize, gps_pos_t &pos, ins_1_t &in
...
-	n += SNPRINTF(a+n, aSize-n, ",%d", INS_STATUS_SOLUTION(ins1.insStatus) >= INS_STATUS_SOLUTION_NAV); //Field 12 - INS Status
+	nmea_sprint(a, aSize, n, ",%d", fix);															// 11 - GPS Quality
+	nmea_sprint(a, aSize, n, ",%d", INS_STATUS_SOLUTION(ins1.insStatus) >= INS_STATUS_SOLUTION_NAV); // 12 - INS Status
 	
-	unsigned int checkSum = ASCII_compute_checksum((uint8_t*)(a+1), n);
-	n += SNPRINTF(a+n, aSize-n, "*%.2x\r\n", checkSum);
+	return nmea_sprint_footer(a, aSize, n);
+}
+
+void print_string_n(char a[], int n)
```
Comment (kylemallory) [inline src/protocol_nmea.cpp:879]: should be in a utility source file.

File: src/protocol_nmea.cpp [hunk 3]
```diff
@@ -851,11 +870,463 @@ int did_gps_to_nmea_pashr(char a[], const int aSize, gps_pos_t &pos, ins_1_t &in
...
+		case '8':	return SAT_SV_SIG_ID_GPS_L5Q;
+		}
+		break;
+    case SAT_SV_GNSS_ID_SBS:
+		return 0;
+    case SAT_SV_GNSS_ID_GAL:
+		switch(nmeaSignalId)
+		{
+		case '7':	return SAT_SV_SIG_ID_Galileo_E1C2;
+		// case '7':	return SAT_SV_SIG_ID_Galileo_E1B2;
```
Comment (kylemallory) [inline src/protocol_nmea.cpp:1028]: Dead code.  Should these be removed?

File: src/test/test_protocol_nmea.cpp [hunk 4]
```diff
@@ -0,0 +1,1283 @@
...
+using namespace std;
+
+
+#define ASCII_BUF_LEN   200
+#define POS_LAT_DEG     40.330578
+#define POS_LON_DEG     -111.725816
+#define POS_ALT_M       1406.39
+#define LEAP_SEC        18
+
+TEST(protocol_nmea, nmea_parse_ascb)
```
Comment (kylemallory) [inline src/test/test_protocol_nmea.cpp:14]: We should also start including a doc block for each of our tests, which describes what the tests do.  Not a requirement, just a suggestion.  It would help all of us to better understand what the individual tests aims to accomplish, and help to determine when it fails, if it failed to catch an edge case, or if there is bigger issue in the test.

File: src/test/test_protocol_nmea.cpp [hunk 5]
```diff
@@ -0,0 +1,1283 @@
...
+
+    // cout << "Before (" << buf.size() << "):\n" << buf;
+    // cout << "After (" << abuf_n << "):\n" << abuf;
+
+    ASSERT_TRUE( abuf_n == buf.size() );
+    ASSERT_TRUE( memcmp(abuf, buf.c_str(), abuf_n) == 0 );
+}
+
+
+#if 1
```
Comment (kylemallory) [inline src/test/test_protocol_nmea.cpp:447]: Remove "ALWAYS TRUE" macro conditions.

File: src/test/test_protocol_nmea.cpp [hunk 6]
```diff
@@ -0,0 +1,1283 @@
...
+TEST(protocol_nmea, binary_GSV_binary)
+{
+    gps_sat_t gpsSat = {};
+    gpsSat.numSats = 33;
+    gpsSat.timeOfWeekMs = 436693200;
+    gps_sat_sv_t *sat = &(gpsSat.sat[0]);
+    sat->azim = 310;
```
Comment (kylemallory) [inline src/test/test_protocol_nmea.cpp:454]: These are a lot of magic numbers; I suspect these are coming from some observations to test against.  Is there a better way to setup this state, so its more apparent what is happening (or what its aiming to test)?

File: src/test/test_protocol_nmea.cpp [hunk 7]
```diff
@@ -0,0 +1,1283 @@
...
+            ASSERT_TRUE( dst.sigId == src.sigId );
+            ASSERT_TRUE( dst.quality == src.quality );
+            ASSERT_TRUE( dst.cno == src.cno );
+        }
+    }
+}
+#endif
+
+
+#if 0   // Used for user manual
```
Comment (kylemallory) [inline src/test/test_protocol_nmea.cpp:1267]: Dead code?  Comment suggests that its for the user manual?  How so?

**Review by kylemallory — COMMENTED** [review]
Big PR.  A lot of my comments are superficial.  Feel free to ignore them, but we do need to get in a better habit of keeping our code well documented (particularly as we anticipate new eyes like mine turning towards it), and clear.

There are a handful that probably should be addressed, but I'll leave it to you.  I know we're under a pinch to get this merged and out.
